### ISSUE TITLE
Bug: BLE for curtains not working after firmware upgrade

### ISSUE THREAD

**Issue by nyjklein** [OpenWonderLabs/homebridge-switchbot#265]
### Describe The Bug

Since a firmware upgrade to the hub and the curtains a couple of days ago, BLE seems to have stopped working. Also, batterystatus no longer seems to be returned. New curtain firmware v3.9. New hub firmware V3.6-2.3.

### To Reproduce

Update firmware

### Expected behavior

No errors

### Relevant log output

```shell
[16/02/2022, 11:58:03] [HB Supervisor] Restarting Homebridge...
[16/02/2022, 11:58:03] [HB Supervisor] Starting Homebridge with extra flags: -I
[16/02/2022, 11:58:03] [HB Supervisor] Started Homebridge v1.4.0 with PID: 4571
[16/02/2022, 11:58:05] Loaded config.json with 9 accessories and 8 platforms.
[16/02/2022, 11:58:05] Loaded 26 cached accessories from cachedAccessories.
[16/02/2022, 11:58:05] ---
[16/02/2022, 11:58:07] Loaded plugin: homebridge-bond@3.2.9
[16/02/2022, 11:58:08] Registering platform 'homebridge-bond.Bond'
[16/02/2022, 11:58:08] ---
[16/02/2022, 11:58:08] Loaded plugin: homebridge-cmdtrigger@0.3.4
[16/02/2022, 11:58:08] Registering accessory 'homebridge-cmdtrigger.CmdTrigger'
[16/02/2022, 11:58:08] ---
[16/02/2022, 11:58:08] Loaded plugin: homebridge-config-ui-x@4.42.0
[16/02/2022, 11:58:08] Registering platform 'homebridge-config-ui-x.config'
[16/02/2022, 11:58:08] ---
[16/02/2022, 11:58:08] Loaded plugin: homebridge-delay-switch@3.0.1
[16/02/2022, 11:58:08] Registering accessory 'homebridge-delay-switch.DelaySwitch'
[16/02/2022, 11:58:08] ---
[16/02/2022, 11:58:08] Loaded plugin: homebridge-dyson-pure-cool@2.4.1
[16/02/2022, 11:58:08] Registering platform 'homebridge-dyson-pure-cool.DysonPureCoolPlatform'
[16/02/2022, 11:58:08] ---
[16/02/2022, 11:58:12] Loaded plugin: homebridge-hue@0.13.38
[16/02/2022, 11:58:12] Registering platform 'homebridge-hue.Hue'
[16/02/2022, 11:58:12] ---
[16/02/2022, 11:58:12] Loaded plugin: homebridge-myq@2.7.4
[16/02/2022, 11:58:12] Registering platform 'homebridge-myq.myQ'
[16/02/2022, 11:58:12] ---
[16/02/2022, 11:58:13] Loaded plugin: homebridge-nest@4.6.4
[16/02/2022, 11:58:13] Registering platform 'homebridge-nest.Nest'
[16/02/2022, 11:58:13] ---
[16/02/2022, 11:58:13] Loaded plugin: homebridge-tplink-smarthome@7.0.1
[16/02/2022, 11:58:13] Registering platform 'homebridge-tplink-smarthome.TplinkSmarthome'
[16/02/2022, 11:58:13] ---
[16/02/2022, 11:58:14] Loaded plugin: @switchbot/homebridge-switchbot@1.12.5
[16/02/2022, 11:58:14] Registering platform '@switchbot/homebridge-switchbot.SwitchBot'
[16/02/2022, 11:58:14] ---
[16/02/2022, 11:58:14] Loading 8 platforms...
[16/02/2022, 11:58:14] [Config] Initializing config platform...
[16/02/2022, 11:58:14] [Config] Running in Service Mode
[16/02/2022, 11:58:14] [SwitchBot] Initializing SwitchBot platform...
[16/02/2022, 11:58:14] [SwitchBot] Platform Config: {"logging":"standard","refreshRate":120}
[16/02/2022, 11:58:16] [SwitchBot] Total SwitchBot Devices Found: 7
[16/02/2022, 11:58:16] [SwitchBot] Restoring existing accessory from cache: Humidifier EE DeviceID: 98F4AB2F20EE
[16/02/2022, 11:58:16] [SwitchBot] Restoring existing accessory from cache: Living Room Curtain DeviceID: DFC80D28CE14
[16/02/2022, 11:58:16] [SwitchBot] Restoring existing accessory from cache: Office Meter DeviceID: E0C8750159BB
[16/02/2022, 11:58:16] [SwitchBot] Restoring existing accessory from cache: Living Room Meter DeviceID: F1BAF4B4E58E
[16/02/2022, 11:58:16] [SwitchBot] Restoring existing accessory from cache: Bedroom Meter DeviceID: F3745EADBBFE
[16/02/2022, 11:58:16] [SwitchBot] Humidifier: Bedroom Humidifier Config: {"hide_temperature":true,"set_minStep":33,"ble":false,"logging":"standard"}
[16/02/2022, 11:58:16] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Using Device Config Logging: debug
[16/02/2022, 11:58:16] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Using Device Config refreshRate: 120
[16/02/2022, 11:58:16] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Using Default Curtain Refresh Rate.
[16/02/2022, 11:58:16] [SwitchBot] Curtain: Living Room Curtain scanDuration is less then updateRate, overriding scanDuration with updateRate
[16/02/2022, 11:58:16] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Using Default scanDuration: 7
[16/02/2022, 11:58:16] [SwitchBot] Curtain: Living Room Curtain Config: {"set_minStep":10,"set_min":5,"set_max":96,"ble":true,"logging":"debug","refreshRate":120}
[16/02/2022, 11:58:16] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BLE refreshStatus
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Add Light Sensor Service
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Add Battery Service
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BLE Address: df:c8:0d:28:ce:14
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentPosition: 0
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic PositionState: 2
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic TargetPosition: 0
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentAmbientLightLevel: undefined
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BatteryLevel: undefined
[16/02/2022, 11:58:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain StatusLowBattery: undefined
[16/02/2022, 11:58:17] Homebridge v1.4.0 (HAP v0.10.0) (Homebridge BFAB) is running on port 51567.
[16/02/2022, 11:58:24] [SwitchBot] Curtain: Living Room Curtain wasn't able to establish BLE Connection
[16/02/2022, 11:58:24] [SwitchBot] Curtain: Living Room Curtain Using OpenAPI Connection
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain OpenAPI refreshStatus
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain refreshStatus: {"statusCode":100,"body":{"deviceId":"DFC80D28CE14","deviceType":"Curtain","hubDeviceId":"DF7B6DA19FE7","calibrate":true,"group":false,"moving":false,"slidePosition":1},"message":"success"}
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain OpenAPI parseStatus
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain Living Room Curtain CurrentPosition: 100
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Standby, CurrentPosition: 100
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Stopped
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentPosition: 100, TargetPosition: 100, PositionState: 2,
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentAmbientLightLevel: 6001
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentPosition: 100
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic PositionState: 2
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic TargetPosition: 100
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentAmbientLightLevel: 6001
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BatteryLevel: undefined
[16/02/2022, 11:58:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain StatusLowBattery: undefined
[16/02/2022, 12:00:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BLE refreshStatus
[16/02/2022, 12:00:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BLE Address: df:c8:0d:28:ce:14
[16/02/2022, 12:00:24] [SwitchBot] Curtain: Living Room Curtain wasn't able to establish BLE Connection
[16/02/2022, 12:00:24] [SwitchBot] Curtain: Living Room Curtain Using OpenAPI Connection
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain OpenAPI refreshStatus
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain refreshStatus: {"statusCode":100,"body":{"deviceId":"DFC80D28CE14","deviceType":"Curtain","hubDeviceId":"DF7B6DA19FE7","calibrate":true,"group":false,"moving":false,"slidePosition":1},"message":"success"}
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain OpenAPI parseStatus
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain Living Room Curtain CurrentPosition: 100
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Standby, CurrentPosition: 100
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Stopped
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentPosition: 100, TargetPosition: 100, PositionState: 2,
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentAmbientLightLevel: 6001
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentPosition: 100
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic PositionState: 2
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic TargetPosition: 100
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentAmbientLightLevel: 6001
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BatteryLevel: undefined
[16/02/2022, 12:00:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain StatusLowBattery: undefined
[16/02/2022, 12:02:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BLE refreshStatus
[16/02/2022, 12:02:17] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BLE Address: df:c8:0d:28:ce:14
[16/02/2022, 12:02:24] [SwitchBot] Curtain: Living Room Curtain wasn't able to establish BLE Connection
[16/02/2022, 12:02:24] [SwitchBot] Curtain: Living Room Curtain Using OpenAPI Connection
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain OpenAPI refreshStatus
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain refreshStatus: {"statusCode":100,"body":{"deviceId":"DFC80D28CE14","deviceType":"Curtain","hubDeviceId":"DF7B6DA19FE7","calibrate":true,"group":false,"moving":false,"slidePosition":97},"message":"success"}
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain OpenAPI parseStatus
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain Living Room Curtain CurrentPosition: 0
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Standby, CurrentPosition: 0
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain Stopped
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentPosition: 0, TargetPosition: 0, PositionState: 2,
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain CurrentAmbientLightLevel: 6001
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentPosition: 0
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic PositionState: 2
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic TargetPosition: 0
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain updateCharacteristic CurrentAmbientLightLevel: 6001
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain BatteryLevel: undefined
[16/02/2022, 12:02:24] [SwitchBot] [DEBUG] Curtain: Living Room Curtain StatusLowBattery: undefined
```

### Config for homebridge-switchbot

```shell
{
    "bridge": {
        "name": "Homebridge BFAB",
        "port": 51567,
    },
    "accessories": [
        {
            "accessory": "CmdTrigger",
            "name": "Hello World",
            "command": "echo Hello World",
            "delay": "10000",
            "execAfterDelay": false
        },
        {
            "accessory": "CmdTrigger",
            "name": "HB Restart",
            "command": "sudo -n systemctl restart homebridge",
            "delay": "10000",
            "execAfterDelay": false
        },
        {
            "name": "Pseudo Hallway",
            "delay": 300000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        },
        {
            "name": "Pseudo 1 Minute",
            "delay": 60000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        },
        {
            "name": "Pseudo 10 Minute",
            "delay": 600000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        },
        {
            "name": "Pseudo 5 Second",
            "delay": 5000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        },
        {
            "name": "Psuedo Enytryway",
            "delay": 300000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        },
        {
            "name": "Pseudo Kitchen",
            "delay": 480000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        },
        {
            "name": "Pseudo Bedroom",
            "delay": 360000,
            "disableSensor": false,
            "startOnReboot": false,
            "accessory": "DelaySwitch"
        }
    ],
    "platforms": [
        {
            "name": "Config",
            "port": 8581,
            "auth": "form",
            "theme": "dark-mode",
            "tempUnits": "f",
            "lang": "auto",
            "platform": "config"
        },
        {
            "name": "Nest",
            "options": [
                "Thermostat.SeparateBuiltInTemperatureSensor.Enable",
                "HomeAway.Disable",
                "Thermostat.Fan.Disable",
                "Thermostat.Eco.Disable"
            ],
            "platform": "Nest"
        },
        {
            "bonds": [
                {
                    "ip_address": "192.168.23.170",
                }
            ],
            "include_dimmer": false,
            "include_toggle_state": false,
            "fan_speed_values": false,
            "platform": "Bond"
        },
        {
            "refreshInterval": 12,
            "activeRefreshInterval": 3,
            "activeRefreshDuration": 300,
            "platform": "myQ"
        },
        {
            "name": "Hue",
            "anyOn": false,
            "effects": true,
            "groups": true,
            "hosts": [
                "192.168.23.165"
            ],
            "nativeHomeKitLights": true,
            "nativeHomeKitSensors": true,
            "resource": true,
            "rooms": true,
            "scenes": true,
            "scenesAsSwitch": false,
            },
            "platform": "Hue"
        },
        {
            "devices": [
                {
                    "ipAddress": "192.168.23.250",
                    "enableAutoModeWhenActivating": true,
                    "enableOscillationWhenActivating": true,
                    "isNightModeEnabled": false,
                    "isJetFocusEnabled": false,
                    "isContinuousMonitoringEnabled": false,
                    "isTemperatureSensorEnabled": false,
                    "isHumiditySensorEnabled": false,
                    "isAirQualitySensorEnabled": false,
                    "isSingleSensorAccessoryModeEnabled": false,
                    "isSingleAccessoryModeEnabled": false,
                    "isFullRangeHumidity": false,
                    "isHeatingDisabled": false
                },
                {
                    "ipAddress": "192.168.23.220",
                    "serialNumber": "E4D-US-NFA0784A",
                },
                {
                    "ipAddress": "192.168.23.246",
                    "serialNumber": "SY7-US-KHA0248A",
                    "enableAutoModeWhenActivating": true,
                    "enableOscillationWhenActivating": true,
                    "isNightModeEnabled": false,
                    "isJetFocusEnabled": false,
                    "isContinuousMonitoringEnabled": false,
                    "isTemperatureSensorEnabled": false,
                    "isHumiditySensorEnabled": false,
                    "isAirQualitySensorEnabled": false,
                    "isSingleSensorAccessoryModeEnabled": false,
                    "isSingleAccessoryModeEnabled": false,
                    "isFullRangeHumidity": false,
                    "isHeatingDisabled": false
                },
                {
                    "ipAddress": "192.168.23.186",
                    "serialNumber": "K1N-US-PHA0456A",
                    "enableAutoModeWhenActivating": true,
                    "enableOscillationWhenActivating": true,
                    "isNightModeEnabled": false,
                    "isJetFocusEnabled": false,
                    "isContinuousMonitoringEnabled": false,
                    "isTemperatureSensorEnabled": true,
                    "isHumiditySensorEnabled": false,
                    "isAirQualitySensorEnabled": false,
                    "isSingleSensorAccessoryModeEnabled": false,
                    "isSingleAccessoryModeEnabled": false,
                    "isFullRangeHumidity": false,
                    "isHeatingDisabled": false
                }
            ],
            "platform": "DysonPureCoolPlatform"
        },
        {
            "name": "TplinkSmarthome",
            "addCustomCharacteristics": true,
            "deviceTypes": [
                "plug",
                "bulb"
            ],
            "platform": "TplinkSmarthome"
        },
        {
            "name": "SwitchBot",
            "credentials": {
                "openToken":                 "notice": "Keep your token a secret!"
            },
            "options": {
                "devices": [
                    {
                        "deviceId": "DFC80D28CE14",
                        "configDeviceName": "Living Room Curtain",
                        "hide_device": false,
                        "configDeviceType": "Curtain",
                        "ble": true,
                        "curtain": {
                            "set_minStep": 10,
                            "set_min": 5,
                            "set_max": 96
                        },
                        "refreshRate": 120,
                        "logging": "debug"
                    },
                    {
                        "deviceId": "98F4AB2F20EE",
                        "configDeviceName": "Bedroom Humidifier",
                        "hide_device": false,
                        "configDeviceType": "Humidifier",
                        "ble": false,
                        "humidifier": {
                            "hide_temperature": true,
                            "set_minStep": 33
                        },
                        "logging": "standard"
                    }
                ],
                "refreshRate": 120,
                "logging": "standard"
            },
            "platform": "SwitchBot"
        }
    ]
}
```

### Device & Model

Raspbian GNU/Linux Buster (10)

### Node.js Version

v16.14.0

### NPM Version

v8.3.1

### Homebridge Version

v1.4.0

### Homebridge Switchbot Plugin Version

v1.12.5

### Homebridge Config UI X Plugin Version

v4.42.0

### Operating System

Raspian

**Comment by donavanbecker** [comment]
@nyjklein Yeah I am having the same issue

**Comment by nyjklein** [comment]
BTW, here's another entry that only shows up when "standard" logging is set instead of "debug":

[16/02/2022, 00:05:21] [@switchbot/homebridge-switchbot] This plugin generated a warning from the characteristic 'Battery Level': characteristic value expected valid finite number and received "NaN" (number). See https://git.io/JtMGR for more info.

**Comment by donavanbecker** [comment]
yes that will happen if bluetooth cannot connect.

**Comment by nyjklein** [comment]
Any prospects here? For now, I turned off BLE for the curtains. That eliminates the delay I was seeing initiating movement (I guess while BLE timed out).

**Comment by donavanbecker** [comment]
Until node-switchbot is fixed to work with the latest firmware. There is nothing I can do.

**Comment by donavanbecker** [comment]
@nyjklein this is fixed in v1.12.6
